### PR TITLE
Add reset progress option to user settings

### LINKED ISSUE #10806: Add "Reset all my progress" button to settings
One of the most popular requests I receive is to reset a camper's progress. Often they come back after a long hiatus and just want to start from scratch. Even though I explain to them that they can just complete the challenges again, most campers still want to reset it. 

the settings page with the added button

We should add this button to the "danger zone" section. It should trigger a warning modal similar to the "delete my account" action. Here's what the copy should say:

the modal with the confirmation text written below

You don't really want to reset your progress, do you?
This will really delete all of your progress and brownie points.
We won't be able to recover any of it for you later, even if you change your mind.
[ Nevermind, I don't want to delete all of my progress and brownie points ]
[ I am 100% sure I want to reset all of my progress and brownie points ]

### PR DESCRIPTION
**Pull request by dhcodes**
#### Pre-Submission Checklist

- [X] Your pull request targets the `staging` branch of FreeCodeCamp.
- [X] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [X] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [X] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

- [X] Tested changes locally.
- [X] Closes currently open issue (replace XXXX with an issue no): Closes #10806
#### Description

Made a new jade page for progress reset, updated settings to add new button to reset progress, wrote functions to reset necessary attributes for user and update progressTimestamp to when the progress is reset. I tested this locally by making a new user, checking the mongo record, completing a few challenges, resetting the progress, and then confirming that the reset record matched the original new record.

### CODE REVIEW

**Comment by dhcodes** [comment]
Will update this based on the Travis build fail

**Comment by dhcodes** [comment]
As an aside, I'm not sure this removed the code stored in the editor from challenges I'd completed with my test user. Is #10200 implemented on `staging`?

File: server/boot/user.js [hunk 1]
```diff
@@ -450,6 +465,33 @@ module.exports = function(app) {
...
+      if (err) { return next(err); }
+      return user.updateAttributes({
+        progressTimestamps: [{
+          timestamp: Date.now()
+        }],
+        currentStreak: 0,
+        longestStreak: 0,
+        currentChallengeId: '',
+        isBackEndCert: false,
+        isFullStackCert: false,
```
Comment (Bouncey) [inline server/boot/user.js]: Is there a 'isFrontEndCert' and/or 'isDataVizCert' that would need reset as well? Or something like that?

File: server/views/account/reset-progress.jade [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        .modal-dialog
+            .modal-content
+                .modal-header
+                    a.close(href='/about', data-dismiss='modal', aria-hidden='true') ×
+                    h3 You don't really want to reset your progress, do you?
+                .modal-body
+                    p This will really delete all of your progress and brownie points.
+                    p We won't be able to recover any of it for you later, even if you change your mind.
+                .modal-footer
+                    a.btn.btn-success.btn-block(href='/about', data-dismiss='modal', aria-hidden='true')
```
Comment (Bouncey) [inline server/views/account/reset-progress.jade]: Why are we sending them to the about page here?

File: server/views/account/reset-progress.jade [hunk 3]
```diff
@@ -0,0 +1,25 @@
+extends ../layout
+block content
+    include ../partials/flyer
+    #modal-dialog.modal.animated.wobble
+        .modal-dialog
+            .modal-content
+                .modal-header
+                    a.close(href='/about', data-dismiss='modal', aria-hidden='true') ×
```
Comment (Bouncey) [inline server/views/account/reset-progress.jade]: Why are we sending them to the about page here?

**Review by Bouncey — COMMENTED** [review]
I will pull down today and test locally

File: server/boot/user.js (hunk 1 above)
Comment (dhcodes) [inline server/boot/user.js]: Good question. I'm slightly confused by how the FrontEnd and DataViz certificates are passed to the user model. They aren't present on the model by default. I think I will add an if statement to check for them.

File: server/views/account/reset-progress.jade (hunk 3 above)
Comment (dhcodes) [inline server/views/account/reset-progress.jade]: Good point, I will send them to `/settings`

File: server/views/account/reset-progress.jade (hunk 2 above)
Comment (dhcodes) [inline server/views/account/reset-progress.jade]: Same. I was modeling the delete user function and was on auto-pilot.

File: server/boot/user.js [hunk 4]
```diff
@@ -183,6 +183,21 @@ module.exports = function(app) {
...
+    sendNonUserToMap,
+    showResetProgress
+  );
+  api.post(
+    '/account/resetprogress',
+    ifNoUser401,
+    postResetProgress
+  );
+  api.get(
+    '/account',
```
Comment (BerkeleyTrue) [inline server/boot/user.js]: You've duplicated the /account logic

File: server/boot/user.js (hunk 1 above)
Comment (BerkeleyTrue) [inline server/boot/user.js]: Yes they should all be set to false.

**Comment by dhcodes** [comment]
I've updated this but can't get the changes to update on the PR. I think I'll have to make a new branch and try again :(

**Comment by dhcodes** [comment]
@Bouncey @BerkeleyTrue I updated based on your feedback and re-tested locally. Should be better now.

File: server/boot/user.js (hunk 1 above)
Comment (BerkeleyTrue) [inline server/boot/user.js]: @dhcodes The two other certs needed to be added here

Comment (dhcodes) [inline server/boot/user.js]: They are, they are just below these github comment boxes.

Comment (BerkeleyTrue) [inline server/boot/user.js]: Oh I see. Thanks

**Comment by BerkeleyTrue** [comment]
LGTM. Can someone test

**Review by thanhjack9x — APPROVED** [review]

**Comment by dhcodes** [comment]
I think this should be good to go.

**Comment by atjonathan** [comment]
Cannot QA. `/settings` automatically redirects to `/map/`. Not sure of the cause, but I don't believe it's an issue with this PR.

cc/ @FreeCodeCamp/issue-moderators

**Comment by ivantedja** [comment]
@atjonathan the redirection probably related with #10879

**Comment by atjonathan** [comment]
@Bouncey, please rebase this off `staging`. The `/settings` route works on the `staging` branch 🎉

**Comment by dhcodes** [comment]
@atjonathan you should be able to get to the settings by clicking your avatar from the map page. Or something. Just keep clicking your avatar and it should get you there eventually. There's an open issue about the map redirect

**Comment by atjonathan** [comment]
@dhcodes, couldn't get that to work on this PR. Works fine on `staging`.

**Comment by ivantedja** [comment]
@atjonathan I don't see any different behaviour between this PR and staging
- from `/map`
- click avatar and it will go to settings

I did reinstall everything though (`node_modules`, `bower_components`, drop database and recreate it)

**Comment by atjonathan** [comment]
@ivantedja, can't get that to work on this PR branch. I will try again

File: server/views/account/reset-progress.jade [hunk 5]
```diff
@@ -0,0 +1,25 @@
+extends ../layout
+block content
+    include ../partials/flyer
+    #modal-dialog.modal.animated.wobble
```
Comment (atjonathan) [inline server/views/account/reset-progress.jade]: @dhcodes, I'm not a fan of this wobble!

File: server/views/account/reset-progress.jade [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        .modal-dialog
+            .modal-content
+                .modal-header
+                    a.close(href='/settings', data-dismiss='modal', aria-hidden='true') ×
+                    h3 You don't really want to reset your progress, do you?
+                .modal-body
+                    p This will really delete all of your progress and brownie points.
+                    p We won't be able to recover any of it for you later, even if you change your mind.
+                .modal-footer
+                    a.btn.btn-success.btn-block(href='/settings', data-dismiss='modal', aria-hidden='true')
```
Comment (atjonathan) [inline server/views/account/reset-progress.jade]: I don't like the color of this, a green might be better?

File: server/views/account/reset-progress.jade [hunk 7]
```diff
@@ -0,0 +1,25 @@
+extends ../layout
+block content
+    include ../partials/flyer
+    #modal-dialog.modal.animated.wobble
+        .modal-dialog
+            .modal-content
```
Comment (atjonathan) [inline server/views/account/reset-progress.jade]: [https://gyazo.com/ba901b0f18c386912c82aeedba5ca9e8](https://gyazo.com/ba901b0f18c386912c82aeedba5ca9e8)

Should this be moved into the center of the screen?
It defiantly shouldn't be overlapping the header.

**Review by ghost — DISMISSED** [review]

**Comment by atjonathan** [comment]
@ivantedja, as I mentioned earlier, the problem was this PR had an earlier version of `staging`. The PR will take a snapshot of what `staging` has after making a branch out of it.

**Comment by dhcodes** [comment]
@atjonathan I replicated the actions and modal of the delete account route so while I'm not opposed to making the changes, we should probably make them to both routes.

**Comment by Bouncey** [comment]
I'm all for removing the animations, they are so '2015'

I will pull this down again today to test

**Comment by dhcodes** [comment]
Should we open a new issue to make the UI changes since they are unrelated to the functionality of this PR? Assuming it tests okay for @Bouncey

**Comment by raisedadead** [comment]
@dhcodes yes that sounds right, please open a new request for the UI changes. 
We can push this if it's good.

**Review by ghost — APPROVED** [review]
#### LGTM :+1:

Everything acts as it's supposed to, I will open a PR with the styling/animation changes soon.
Thanks @dhcodes ❤️ 💪

**Pull request merged**